sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedDeltaLengthByteArrayReader.java

```diff
@@ -0,0 +1,103 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.parquet;
+
+import static org.apache.spark.sql.types.DataTypes.IntegerType;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.apache.spark.memory.MemoryMode;
+import org.apache.spark.sql.execution.vectorized.OffHeapColumnVector;
+import org.apache.spark.sql.execution.vectorized.OnHeapColumnVector;
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+
+/**
+ * An implementation of the Parquet DELTA_LENGTH_BYTE_ARRAY decoder that supports the vectorized
+ * interface.
+ */
+public class VectorizedDeltaLengthByteArrayReader extends VectorizedReaderBase implements
+    VectorizedValuesReader {
+
+  private final MemoryMode memoryMode;
+  private final VectorizedDeltaBinaryPackedReader lengthReader =
+      new VectorizedDeltaBinaryPackedReader();
+  private ByteBufferInputStream in;
+  private WritableColumnVector lengthsVector;
+  private int currentRow = 0;
+
+  VectorizedDeltaLengthByteArrayReader(MemoryMode memoryMode) {
+    this.memoryMode = memoryMode;
+  }
+
+  @Override
+  public void initFromPage(int valueCount, ByteBufferInputStream in) throws IOException {
+    if (memoryMode == MemoryMode.OFF_HEAP) {
+      lengthsVector = new OffHeapColumnVector(valueCount, IntegerType);
+      lengthsVector.putInts(0, valueCount, 0);
+    } else {
+      lengthsVector = new OnHeapColumnVector(valueCount, IntegerType);
+    }
+    lengthReader.initFromPage(valueCount, in);
+    lengthReader.readIntegers(lengthReader.getTotalValueCount(), lengthsVector, 0);
+    this.in = in.remainingStream();
+  }
+
+  @Override
+  public void readBinary(int total, WritableColumnVector c, int rowId) {
+    if (total == 0) {
+      return;
+    }
+    ByteBuffer buffer;
+    ByteBufferOutputWriter outputWriter;
+    if (memoryMode == MemoryMode.OFF_HEAP) {
+      outputWriter = ByteBufferOutputWriter::copyWriteByteBuffer;
+    } else {
+      outputWriter = ByteBufferOutputWriter::writeArrayByteBuffer;
+    }
+    int length;
+    for (int i = 0; i < total; i++) {
+      length = lengthsVector.getInt(rowId + i);
```

Comment:
if `length` is 0, can we simplify the operation of `line77 ~ line82`

Reply:
I'm not entirely sure. Does the WritableColumnVector allow skipping values? I assume so, but I'm not sure. Do you know?

Reply:
null values will not be handled here since they are represented using definition levels in Parquet. I think only empty strings matter here but it's likely not so common.

Reply:
Right, I missed that @LuciferYang was talking about null values. This method gets called only for non-null values.
I meant that if the ith value is an empty string and we simply do not write the ith value to the vector, does the vector guarantee that the ith value will be treated as an empty string? 
If so we can simply not write the value if the length is zero. We'll save a couple of function calls for empty strings but we will add an if condition for every value. Probably not worth it.

Reply:
@parthchandra  if `This method gets called only for non-null values.`,  then there should be no need to add `lengthsVector.putInts(0, valueCount, 0)` because the place where the exception was thrown was originally a null value

Reply:
It seems that there are many strings with length of zero when `String with Nulls Scan` is running. I'm not sure whether there is a problem with the design of `String with Nulls Scan`.

Reply:
Or `lengthsVector ` can't distinguish between zero-length string and Null?

Reply:
If the column encoding is `DELTA_BYTE_ARRAY` the data is in encoded as a vector of prefix lengths and a vector of suffixes. The vector of suffixes is a `DELTA_LENGTH_BYTE_ARRAY` encoding. So we may have a `DELTA_LENGTH_BYTE_ARRAY` where we have empty strings even if the value is not null. (AFAIK this is the only use of `DELTA_LENGTH_BYTE_ARRAY`). 
Also, `lengthsVector` as you noted, cannot distinguish between empty string and null bin `readBinary` scope because it doesn't have the definition level.

Reply:
So I think it may be valuable to avoid some unnecessary operations when length is 0. At the same time, this may be related to the performance degradation in `String with Nulls Scan(95%)` scenario, because 95% of the values in this scenario are values with length of 0.

Reply:
Added the check. Locally run benchmark showed no real gain though.

Reply:
I run the benchmark several times today use GA,  It seems that the new pr in `String with Nulls Scan(95%)` scene is still slower than before. 

And I run this case locally(use Apple Silicon), there is really no significant difference in M1.

Also cc @dongjoon-hyun @viirya @sunchao  for further review

Reply:
> 
> Also cc @dongjoon-hyun @viirya @sunchao for further review
@dongjoon-hyun, @viirya, @sunchao any further comments?

Reply:
will take a look soon, sorry missed this ping.

Reply:
Hmm I think we should not skip writing if `length` is 0. Maybe:
```java
    for (int i = 0; i < total; i++) {
      length = lengthsVector.getInt(rowId + i);
      try {
        buffer = in.slice(length);
      } catch (EOFException e) {
        throw new ParquetDecodingException("Failed to read " + length + " bytes");
      }
      outputWriter.write(c, rowId + i, buffer, length);
    }

```

Reply:
☝️ @LuciferYang , @sunchao, should I remove this or not?

Reply:
We should probably have a test case BTW

Reply:
Done.

Reply:
Got it, agree with @sunchao